mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -15,60 +16,106 @@ use std::collections::HashMap;
 use super::EthCoin;
 use crate::{eth::{u256_to_big_decimal, Erc20TokenInfo},
             BalanceError, MmCoin};
+struct BalanceData {
+    ticker: String,
+    address: String,
+    balance: BigDecimal,
+}
+
+struct BalanceFetchError {
+    ticker: String,
+    address: String,
+    error: MmError<BalanceError>,
+}
+
+type BalanceResult = Result<BalanceData, BalanceFetchError>;
 
 /// This implementation differs from others, as they immediately return
 /// an error if any of the requests fails. This one completes all futures
 /// and returns their results individually.
-async fn get_all_balance_results_concurrently(
-    coin: &EthCoin,
-) -> Vec<Result<(String, BigDecimal), (String, MmError<BalanceError>)>> {
+async fn get_all_balance_results_concurrently(coin: &EthCoin) -> Vec<BalanceResult> {
     let mut tokens = coin.get_erc_tokens_infos();
-
     // Workaround for performance purposes.
     //
     // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
     // type and mapping the platform coin and the entire token list (which can grow at any time), we map
     // the platform coin to Erc20TokenInfo so that we can use the token list right away without
     // additional mapping.
     tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
-        token_address: coin.my_address,
+        token_address: Address::default(),
         decimals: coin.decimals,
     });
+    drop_mutability!(tokens);
+
+    let addresses = match coin.my_addresses().await {
+        Ok(addresses) => addresses,
+        Err(e) => {
+            return vec![Err(BalanceFetchError {
+                ticker: coin.ticker.clone(),
+                address: "".to_string(),
+                error: e.into(),
+            })];
+        },
+    };
```

Comment:
I think it's better to just unwrap it if it never fails? Without an address the error doesn't make sense since we propagate the error to the client using both the address and ticker.

Reply:
> I think it's better to just unwrap it if it never fails?

It should never fail, but if we used unwrap, if a dev made an error that is missed in a review it will lead to panicking. I will add an additional error log as you suggested here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979#discussion_r1540476735
I can use `my_addresses` outside `get_all_balance_results_concurrently` instead and pass them to `get_all_balance_results_concurrently`, if `my_addresses` fails we can send the error as an event which will not have address in it.

Reply:
> I can use my_addresses outside get_all_balance_results_concurrently instead and pass them to get_all_balance_results_concurrently, if my_addresses fails we can send the error as an event which will not have address in it.

Fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/846c22a70ab2f62d368908f0e383681679cd1736